### PR TITLE
Enforce that P256Keypair is not copyable

### PR DESCRIPTION
**Pull request by ksperling-apple**
Enforce that `P256Keypair` is not copyable. `P256PublicKey` IS copyable (and actually copied in practice), so prevent
sliced copies by making it final. Also use `= default` instead of empty constructor / destructor bodies.

### CODE REVIEW

File: src/crypto/CHIPCryptoPAL.h
Comment (tehampson) [inline src/crypto/CHIPCryptoPAL.h:492]: quick question, how does this prevent subtypes from either not being copyable or final?

Or is this just a general comment that ends up being enforced by code audits?

Comment (tehampson) [reply src/crypto/CHIPCryptoPAL.h:492]: @tehampson 
^ Just mentioning myself so that I get emails when reply happens on this PR

**Review by tehampson — APPROVED** [review]

File: src/crypto/CHIPCryptoPAL.h
Comment (tcarmelveilleux) [inline src/crypto/CHIPCryptoPAL.h:225]: This breaks the public API. Please stand by while we check if this break our integration. Please wait until after holidays to merge.

**Review by tcarmelveilleux — CHANGES_REQUESTED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
Approved, assuming @tcarmelveilleux is ok with it.

File: src/crypto/CHIPCryptoPAL.h
Comment (ksperling-apple) [reply src/crypto/CHIPCryptoPAL.h:492]: Having them protected means you can't directly use the base type to do a copy, e.g.
```P256KeypairBase copy = someOtherKaypair;```

Each sub-class still has to deal with whether copying should be possible or not.

Comment (ksperling-apple) [reply src/crypto/CHIPCryptoPAL.h:492]: > Or is this just a general comment that ends up being enforced by code audits?

Yes, that's right.

File: src/crypto/CHIPCryptoPAL.h
Comment (ksperling-apple) [reply src/crypto/CHIPCryptoPAL.h:225]: Unless you're using `ECPKey` directly (rather than a sub-class) this shouldn't really change anything, but happy to wait until you've validated this doesn't break things.

File: src/crypto/CHIPCryptoPAL.h
Comment (tehampson) [reply src/crypto/CHIPCryptoPAL.h:492]: > Having them protected means you can't directly use the base type to do a copy, e.g.
`P256KeypairBase copy = someOtherKaypair;`
Each sub-class still has to deal with whether copying should be possible or not.

Thanks for that explanation, very clear now.

**Comment by ksperling-apple** [comment]
Darwin changes landed separately in #31141

File: src/crypto/CHIPCryptoPAL.h
Comment (tehampson) [reply src/crypto/CHIPCryptoPAL.h:225]: @tcarmelveilleux 

Confirmed with Albert that this will only break tests that return a copy, but mentioned that can work around that when we roll

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**
